qiskit/compiler/transpiler.py

```diff
@@ -58,7 +59,8 @@ def transpile(
     dt: Optional[float] = None,
     approximation_degree: Optional[float] = None,
     seed_transpiler: Optional[int] = None,
-    optimization_level: Optional[int] = None,
+    inst_map: Dict[str, Dict[Tuple[int], Schedule]] = None,
+    optimization_level: Union[Optional[int], Optional[str]] = None,
```

Comment:
For the algorithms, that create circuits and execute them, this is presently done via a QuantumInstance that wraps a backend and has common circuit support for the algorithms, that also allows an end user to set circuit aspects such as transpiler options. See here https://github.com/Qiskit/qiskit-terra/blob/3bb72d340fcf9552d4a6bd54b675399578c726db/qiskit/utils/quantum_instance.py#L71
So if this is introduced as a string I think the above should follow suit and be changed likewise. As the code just stores and passes this on I think its a minor change there just for the typehint.

As a very minor comment I think the typehint can be written more simply as `Optional[Union[int, str]] = None`

Reply:
I actually think a better interface for this is the `FullPassManager` being introduced in: https://github.com/Qiskit/qiskit-terra/pull/6403 with for the extra passes the user just builds a post-optimization pass manager that runs these additional passes and sets that on level 3. I don't think we want to use optimization level like this.